File: telegram_validator.py
#!/usr/bin/env python3
"""
Validador de Token do Telegram - Solução Alternativa
Este script valida tokens do Telegram diretamente com a API oficial
"""

import requests
import json
import sys
from datetime import datetime

class TelegramValidator:
    def __init__(self):
        self.base_url = "https://api.telegram.org/bot"
    
    def validate_token(self, token):
        """
        Valida um token do Telegram fazendo uma requisição para getMe
        """
        try:
            url = f"{self.base_url}{token}/getMe"
            response = requests.get(url, timeout=10)
            
            if response.status_code == 200:
                data = response.json()
                if data.get('ok'):
                    bot_info = data.get('result', {})
                    return {
                        'success': True,
                        'message': 'Token válido',
                        'bot_info': {
                            'id': bot_info.get('id'),
                            'is_bot': bot_info.get('is_bot'),
                            'first_name': bot_info.get('first_name'),
                            'username': bot_info.get('username'),
                            'can_join_groups': bot_info.get('can_join_groups'),
                            'can_read_all_group_messages': bot_info.get('can_read_all_group_messages'),
                            'supports_inline_queries': bot_info.get('supports_inline_queries')
                        }
                    }
                else:
                    return {
                        'success': False,
                        'message': f"Erro da API: {data.get('description', 'Erro desconhecido')}"
                    }
            else:
                return {
                    'success': False,
                    'message': f"Erro HTTP {response.status_code}: {response.text}"
                }
                
        except requests.exceptions.Timeout:
            return {
                'success': False,
                'message': 'Timeout na requisição para a API do Telegram'
            }
        except requests.exceptions.ConnectionError:
            return {
                'success': False,
                'message': 'Erro de conexão com a API do Telegram'
            }
        except Exception as e:
            return {
                'success': False,
                'message': f'Erro inesperado: {str(e)}'
            }
    
    def test_send_message(self, token, chat_id, message):
        """
        Testa o envio de uma mensagem usando o token
        """
        try:
            url = f"{self.base_url}{token}/sendMessage"
            payload = {
                'chat_id': chat_id,
                'text': message
            }
            
            response = requests.post(url, json=payload, timeout=10)
            
            if response.status_code == 200:
                data = response.json()
                if data.get('ok'):
                    return {
                        'success': True,
                        'message': 'Mensagem enviada com sucesso',
                        'message_id': data.get('result', {}).get('message_id')
                    }
                else:
                    return {
                        'success': False,
                        'message': f"Erro ao enviar mensagem: {data.get('description', 'Erro desconhecido')}"
                    }
            else:
                return {
                    'success': False,
                    'message': f"Erro HTTP {response.status_code}: {response.text}"
                }
                
        except Exception as e:
            return {
                'success': False,
                'message': f'Erro ao enviar mensagem: {str(e)}'
            }

def main():
    """Função principal para uso via linha de comando"""
    if len(sys.argv) < 2:
        print("Uso: python telegram_validator.py <token> [chat_id] [mensagem]")
        sys.exit(1)
    
    token = sys.argv[1]
    validator = TelegramValidator()
    
    # Validar token
    print(f"🔍 Validando token: {token[:20]}...")
    result = validator.validate_token(token)
    
    print(f"\n📊 Resultado da validação:")
    print(f"Status: {'✅ SUCESSO' if result['success'] else '❌ ERRO'}")
    print(f"Mensagem: {result['message']}")
    
    if result['success'] and 'bot_info' in result:
        bot_info = result['bot_info']
        print(f"\n🤖 Informações do Bot:")
        print(f"ID: {bot_info['id']}")
        print(f"Nome: {bot_info['first_name']}")
        print(f"Username: @{bot_info['username']}")
        print(f"É Bot: {bot_info['is_bot']}")
        print(f"Pode entrar em grupos: {bot_info['can_join_groups']}")
        print(f"Pode ler mensagens de grupo: {bot_info['can_read_all_group_messages']}")
        print(f"Suporte a consultas inline: {bot_info['supports_inline_queries']}")
        
        # Teste de envio de mensagem se chat_id foi fornecido
        if len(sys.argv) >= 3:
            chat_id = sys.argv[2]
            message = sys.argv[3] if len(sys.argv) >= 4 else f"🧪 Teste de conexão - {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}"
            
            print(f"\n📤 Testando envio de mensagem para chat {chat_id}...")
            send_result = validator.test_send_message(token, chat_id, message)
            print(f"Status: {'✅ SUCESSO' if send_result['success'] else '❌ ERRO'}")
            print(f"Mensagem: {send_result['message']}")
    
    # Retornar código de saída apropriado
    sys.exit(0 if result['success'] else 1)

if __name__ == "__main__":
    main()

File: telegram_validator.js
#!/usr/bin/env node
/**
 * Validador de Token do Telegram - Solução Alternativa
 * Este script valida tokens do Telegram diretamente com a API oficial
 */

import https from 'https';
import http from 'http';
import { fileURLToPath } from 'url';
import { dirname } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

class TelegramValidator {
    constructor() {
        this.baseUrl = 'https://api.telegram.org/bot';
    }

    /**
     * Faz uma requisição HTTP/HTTPS
     */
    makeRequest(url, options = {}) {
        return new Promise((resolve, reject) => {
            const urlObj = new URL(url);
            const isHttps = urlObj.protocol === 'https:';
            const client = isHttps ? https : http;
            
            const requestOptions = {
                hostname: urlObj.hostname,
                port: urlObj.port || (isHttps ? 443 : 80),
                path: urlObj.pathname + urlObj.search,
                method: options.method || 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    'User-Agent': 'TelegramValidator/1.0',
                    ...options.headers
                },
                timeout: 10000
            };

            const req = client.request(requestOptions, (res) => {
                let data = '';
                
                res.on('data', (chunk) => {
                    data += chunk;
                });
                
                res.on('end', () => {
                    try {
                        const jsonData = JSON.parse(data);
                        resolve({
                            statusCode: res.statusCode,
                            data: jsonData,
                            headers: res.headers
                        });
                    } catch (error) {
                        resolve({
                            statusCode: res.statusCode,
                            data: data,
                            headers: res.headers,
                            parseError: error.message
                        });
                    }
                });
            });

            req.on('error', (error) => {
                reject(error);
            });

            req.on('timeout', () => {
                req.destroy();
                reject(new Error('Request timeout'));
            });

            if (options.body) {
                req.write(JSON.stringify(options.body));
            }

            req.end();
        });
    }

    /**
     * Valida um token do Telegram fazendo uma requisição para getMe
     */
    async validateToken(token) {
        try {
            const url = `${this.baseUrl}${token}/getMe`;
            const response = await this.makeRequest(url);
            
            if (response.statusCode === 200 && response.data.ok) {
                const botInfo = response.data.result;
                return {
                    success: true,
                    message: 'Token válido',
                    botInfo: {
                        id: botInfo.id,
                        isBot: botInfo.is_bot,
                        firstName: botInfo.first_name,
                        username: botInfo.username,
                        canJoinGroups: botInfo.can_join_groups,
                        canReadAllGroupMessages: botInfo.can_read_all_group_messages,
                        supportsInlineQueries: botInfo.supports_inline_queries
                    }
                };
            } else {
                return {
                    success: false,
                    message: `Erro da API: ${response.data.description || 'Erro desconhecido'}`,
                    statusCode: response.statusCode
                };
            }
        } catch (error) {
            return {
                success: false,
                message: `Erro de conexão: ${error.message}`,
                error: error.code || error.message
            };
        }
    }

    /**
     * Testa o envio de uma mensagem usando o token
     */
    async testSendMessage(token, chatId, message) {
        try {
            const url = `${this.baseUrl}${token}/sendMessage`;
            const response = await this.makeRequest(url, {
                method: 'POST',
                body: {
                    chat_id: chatId,
                    text: message
                }
            });
            
            if (response.statusCode === 200 && response.data.ok) {
                return {
                    success: true,
                    message: 'Mensagem enviada com sucesso',
                    messageId: response.data.result.message_id
                };
            } else {
                return {
                    success: false,
                    message: `Erro ao enviar mensagem: ${response.data.description || 'Erro desconhecido'}`,
                    statusCode: response.statusCode
                };
            }
        } catch (error) {
            return {
                success: false,
                message: `Erro ao enviar mensagem: ${error.message}`,
                error: error.code || error.message
            };
        }
    }
}

/**
 * Função principal para uso via linha de comando
 */
async function main() {
    const args = process.argv.slice(2);
    
    if (args.length < 1) {
        console.log('Uso: node telegram_validator.js <token> [chat_id] [mensagem]');
        process.exit(1);
    }
    
    const token = args[0];
    const chatId = args[1];
    const message = args[2] || `🧪 Teste de conexão - ${new Date().toLocaleString('pt-BR')}`;
    
    const validator = new TelegramValidator();
    
    // Validar token
    console.log(`🔍 Validando token: ${token.substring(0, 20)}...`);
    const result = await validator.validateToken(token);
    
    console.log(`\n📊 Resultado da validação:`);
    console.log(`Status: ${result.success ? '✅ SUCESSO' : '❌ ERRO'}`);
    console.log(`Mensagem: ${result.message}`);
    
    if (result.success && result.botInfo) {
        const botInfo = result.botInfo;
        console.log(`\n🤖 Informações do Bot:`);
        console.log(`ID: ${botInfo.id}`);
        console.log(`Nome: ${botInfo.firstName}`);
        console.log(`Username: @${botInfo.username}`);
        console.log(`É Bot: ${botInfo.isBot}`);
        console.log(`Pode entrar em grupos: ${botInfo.canJoinGroups}`);
        console.log(`Pode ler mensagens de grupo: ${botInfo.canReadAllGroupMessages}`);
        console.log(`Suporte a consultas inline: ${botInfo.supportsInlineQueries}`);
        
        // Teste de envio de mensagem se chat_id foi fornecido
        if (chatId) {
            console.log(`\n📤 Testando envio de mensagem para chat ${chatId}...`);
            const sendResult = await validator.testSendMessage(token, chatId, message);
            console.log(`Status: ${sendResult.success ? '✅ SUCESSO' : '❌ ERRO'}`);
            console.log(`Mensagem: ${sendResult.message}`);
        }
    }
    
    // Retornar código de saída apropriado
    process.exit(result.success ? 0 : 1);
}

// Executar se chamado diretamente
const isMainModule = import.meta.url.startsWith('file:') && 
                     process.argv[1] && 
                     import.meta.url.includes(process.argv[1].replace(/\\/g, '/'));

if (isMainModule) {
    main().catch(error => {
        console.error('❌ Erro inesperado:', error.message);
        process.exit(1);
    });
}

export default TelegramValidator;